octopusdeploy_framework/datasource_tenant_projects.go

```diff
@@ -0,0 +1,201 @@
+	var data TenantProjectsDataModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tenantIDs := util.ExpandStringList(data.TenantIDs)
+	projectIDs := util.ExpandStringList(data.ProjectIDs)
+	environmentIDs := util.ExpandStringList(data.EnvironmentIDs)
+	spaceID := data.SpaceID.ValueString()
+
+	if tenantIDs == nil && projectIDs == nil {
+		resp.Diagnostics.AddError("must provide at least one tenant or one project", "tenant IDs and project IDs are nil")
+	}
+
+	var tenantData []*tenants.Tenant
+	if tenantIDs == nil {
+		tenantData = make([]*tenants.Tenant, 0)
+		for _, projectID := range projectIDs {
+			// todo: should consider using concurrency
+			tenantsByProjectID, err := getTenantByProjectID(t.Client, projectID, spaceID)
+			if err != nil {
+				resp.Diagnostics.AddError("unable to load tenant data for project", err.Error())
+				return
+			}
+			tenantData = append(tenantData, tenantsByProjectID...)
+		}
+	} else {
+		tenantQuery := tenants.TenantsQuery{
+			IDs: tenantIDs,
+		}
+		tenantResource, err := tenants.Get(t.Client, spaceID, tenantQuery)
+		if err != nil {
+			resp.Diagnostics.AddError("unable to load tenant data", err.Error())
+			return
+		}
+		tenantData, err = tenantResource.GetAllPages(t.Client.Sling())
+		if err != nil {
+			resp.Diagnostics.AddError("unable to load tenant data", err.Error())
+			return
+		}
+	}
+
+	// todo: refactor
+	flattenedTenantProjects := make([]any, 0)
+	for _, tenant := range tenantData {
+		if projectIDs == nil {
+			for projectID, envIDs := range tenant.ProjectEnvironments {
+				if environmentIDs != nil {
+					for _, envID := range envIDs {
+						if slices.Contains(environmentIDs, envID) {
+							flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+							break
+						}
+					}
+				} else {
+					flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+				}
+			}
+			continue
+		}
+		for _, projectID := range projectIDs {
+			if envIDs, ok := tenant.ProjectEnvironments[projectID]; ok {
+				if environmentIDs != nil {
+					for _, envID := range envIDs {
+						if slices.Contains(environmentIDs, envID) {
+							flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+							break
+						}
+					}
+				} else {
+					flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+				}
+			}
+
+		}
+	}
+
+	data.TenantProjects, _ = types.ListValueFrom(ctx, types.ObjectType{AttrTypes: tenantProjectType()}, flattenedTenantProjects)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func getTenantByProjectID(c *client.Client, projectID string, spaceID string) ([]*tenants.Tenant, error) {
+	tenantQuery := tenants.TenantsQuery{
+		ProjectID: projectID,
+	}
+	tenantResource, err := tenants.Get(c, spaceID, tenantQuery)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
If a `ProjectID` is provided that doesn't exist, we don't get any API errors here and so it just silently returns nothing. 

It might be helpful for customers if we could validate the existence of the Project and provide a log message if it doesn't exist rather than just having it quietly fail. Not sure how this matches existing approaches in the provider, or how TF Providers are written in general.

Trade-off: seems like we would need to hit another API endpoint first to do so, which could be expensive.

Reply:
I thought about this. As you, its a bit expensive. having a new endpoint would be ideal.